otherlibs/dune-rpc/private/dune_rpc_private.ml

```diff
@@ -592,13 +594,24 @@ module Client = struct
         ; write : Sexp.t list option -> unit Fiber.t
         ; mutable closed_read : bool
         ; mutable closed_write : bool
+        ; on_close : unit Fiber.Ivar.t
         }
 
       let of_chan c =
-        { read = (fun () -> Chan.read c)
+        let on_close = Fiber.Ivar.create () in
+        let read () =
+          let* result = Chan.read c in
+          match result with
+          | None ->
+            let+ () = Fiber.Ivar.fill on_close () in
+            None
+          | _ -> Fiber.return result
+        in
+        { read
         ; write = (fun s -> Chan.write c s)
         ; closed_read = false
         ; closed_write = false
+        ; on_close
```

Comment:
Could your renamed this to `disconnected`? Or alternatively, rename `disconnected` to `on_close`? I don't see the need to make use of both terms in the code base.

Reply:
Done